### PR TITLE
🚀 Feat/liquidity pools v2 

### PR DESCRIPTION
**Pull request by MTthoas**
LiquidityPools / FactoryPools 

- Initializable
- ReentrancyGuardUpgradeable
- AccessManagerUpgradeable
- ERC20Upgradeable
- IAccessManager

Deployment script with real tests.

### CODE REVIEW

File: contracts/src/LiquidityPoolV2.sol [hunk 1]
```diff
@@ -90,10 +123,21 @@ contract LiquidityPoolV2 is ReentrancyGuardUpgradeable, AccessManagedUpgradeable
         emit LiquidityRemoved(msg.sender, tokenAAmount, tokenBAmount);
```
Comment (Ronfflex) [inline contracts/src/LiquidityPoolV2.sol:123]: adjustForDecimals (see last comment of the file to understand)

Can't comment the good line idk why but replace tokenAAmount and tokenBAmount in the transfers but the adjustedToken

File: contracts/src/LiquidityPoolV2.sol [hunk 2]
```diff
@@ -102,47 +146,60 @@ contract LiquidityPoolV2 is ReentrancyGuardUpgradeable, AccessManagedUpgradeable
         // Transfer the input tokens from the user to the contract
         ERC20Upgradeable(tokenIn).transferFrom(msg.sender, address(this), amountIn);
```
Comment (Ronfflex) [inline contracts/src/LiquidityPoolV2.sol:147]: **adjustForDecimals** (see last comment of the file to understand)

Replace amountIn in the transfers but the adjustedAmountIn

File: contracts/src/LiquidityPoolV2.sol [hunk 3]
```diff
@@ -102,47 +146,60 @@ contract LiquidityPoolV2 is ReentrancyGuardUpgradeable, AccessManagedUpgradeable
...
 
-        // Calculate the output amount using the constant product formula
-        uint256 amountInWithFee = amountIn * (10000 - platformFee) / 10000;
+        // Calculate the output amount using the constant product formula, adjusted for the platform fee
+        uint256 amountInWithFee = (amountIn * (10000 - platformFee)) / 10000;
```
Comment (Ronfflex) [inline contracts/src/LiquidityPoolV2.sol:150]: **adjustForDecimals** (see last comment of the file to understand)

Replace amountIn in the transfers but the adjustedAmountIn

File: contracts/src/LiquidityPoolV2.sol [hunk 4]
```diff
@@ -29,19 +36,41 @@ contract LiquidityPoolV2 is ReentrancyGuardUpgradeable, AccessManagedUpgradeable
...
+        _;
+    }
+
+    /**
+     * @notice Function to add liquidity to the pool.
+     * @param tokenAAmount Amount of tokenA to add.
+     * @param tokenBAmount Amount of tokenB to add.
+     */
+    function addLiquidity(uint256 tokenAAmount, uint256 tokenBAmount) external onlyRole(1) nonReentrant {
+        require(tokenAAmount > 0 && tokenBAmount > 0, 'LiquidityPool: amounts must be greater than zero');
```
Comment (Ronfflex) [inline contracts/src/LiquidityPoolV2.sol:74]: **adjustForDecimals** (see last comment of the file to understand)

```solidity
// Adjust token amounts based on decimals
uint256 adjustedTokenAAmount = adjustForDecimals(tokenA, tokenAAmount);
uint256 adjustedTokenBAmount = adjustForDecimals(tokenB, tokenBAmount);
```

File: contracts/src/LiquidityPoolV2.sol [hunk 5]
```diff
@@ -29,19 +36,41 @@ contract LiquidityPoolV2 is ReentrancyGuardUpgradeable, AccessManagedUpgradeable
...
 
         // Transfer tokens from sender to the contract
         ERC20Upgradeable(tokenA).transferFrom(msg.sender, address(this), tokenAAmount);
```
Comment (Ronfflex) [inline contracts/src/LiquidityPoolV2.sol:76]: **adjustForDecimals** (see last comment of the file to understand)

Replace tokenAAmount and tokenBAmount in the transfers but the adjustedToken

File: contracts/src/LiquidityPoolV2.sol [hunk 6]
```diff
@@ -50,14 +79,15 @@ contract LiquidityPoolV2 is ReentrancyGuardUpgradeable, AccessManagedUpgradeable
         // Calculate liquidity tokens to mint
         uint256 liquidity;
         if (totalLiquidity == 0) {
+            // For the first liquidity provider, mint liquidity tokens 1:1 with the provided tokens
             liquidity = Math.sqrt(tokenAAmount * tokenBAmount);
```
Comment (Ronfflex) [inline contracts/src/LiquidityPoolV2.sol:83]: **adjustForDecimals** (see last comment of the file to understand)

Replace tokenAAmount and tokenBAmount in the transfers but the adjustedToken

File: contracts/src/LiquidityPoolV2.sol [hunk 7]
```diff
@@ -50,14 +79,15 @@ contract LiquidityPoolV2 is ReentrancyGuardUpgradeable, AccessManagedUpgradeable
...
         } else {
-            liquidity = Math.min(
-                (tokenAAmount * totalLiquidity) / reserveA,
-                (tokenBAmount * totalLiquidity) / reserveB
-            );
+            // For subsequent liquidity providers, mint liquidity tokens proportionally
+            uint256 amountA = (tokenAAmount * totalLiquidity) / reserveA;
+            uint256 amountB = (tokenBAmount * totalLiquidity) / reserveB;
```
Comment (Ronfflex) [inline contracts/src/LiquidityPoolV2.sol:87]: **adjustForDecimals** (see last comment of the file to understand)

Replace tokenAAmount and tokenBAmount in the transfers but the adjustedToken

File: .gitmodules [hunk 8]
```diff
@@ -1,12 +1,15 @@
...
+	url = https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable
+[submodule "contracts/lib/openzeppelin-contracts"]
+	path = contracts/lib/openzeppelin-contracts
+	url = https://github.com/OpenZeppelin/openzeppelin-contracts
+[submodule "contracts/lib/openzeppelin-foundry-upgrades"]
+	path = contracts/lib/openzeppelin-foundry-upgrades
+	url = https://github.com/OpenZeppelin/openzeppelin-foundry-upgrades
+[submodule "lib/forge-std"]
+	path = lib/forge-std
+	url = https://github.com/foundry-rs/forge-std
```
Comment (Ronfflex) [inline .gitmodules:15]: Delete that. Wrong path and duplicated.

**Review by Ronfflex — CHANGES_REQUESTED** [review]
We need to find a way to fix the end of line sequence (LF or CRLF) the same for everyone because it's horrible to review. (LF would be nice by default)

File: contracts/src/LiquidityPoolFactoryV2.sol [hunk 9]
```diff
@@ -1,60 +1,87 @@
...
-        return allPools;
+    /**
+     * @notice Modifier to restrict function access to accounts with a specific role.
+     * @param roleId ID of the required role.
+     */
+    modifier onlyRole(uint64 roleId) {
+        (bool hasRole, ) = hasRole(roleId, msg.sender);
+        require(hasRole, 'Factory: sender must have the required role');
+        _;
     }
```
[the remark this answers is not in the document]
Comment (MTthoas) [reply contracts/src/LiquidityPoolFactoryV2.sol:86]: Im not agree with that, create specific functions to handle & understand erros is better

File: contracts/src/LiquidityPoolV2.sol [hunk 10]
```diff
@@ -1,17 +1,24 @@
...
+import '@openzeppelin/contracts/utils/math/Math.sol';
+import '@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import { AccessManagerUpgradeable } from '@openzeppelin/contracts-upgradeable/access/manager/AccessManagerUpgradeable.sol';
+import { IAccessManager } from '@openzeppelin/contracts/access/manager/IAccessManager.sol';
+
+interface IAccessManagerExtended is IAccessManager {
+    function hasRole(uint64 roleId, address account) external view returns (bool, uint32);
+}
```
[the remark this answers is not in the document]
Comment (MTthoas) [reply contracts/src/LiquidityPoolV2.sol:13]: Im using hasRole from IAccessManager

File: contracts/src/LiquidityPoolV2.sol [hunk 11]
```diff
@@ -29,19 +36,41 @@ contract LiquidityPoolV2 is ReentrancyGuardUpgradeable, AccessManagedUpgradeable
...
-        require(tokenAAmount > 0 && tokenBAmount > 0, "LiquidityPool: amounts must be greater than zero");
+    /**
+     * @notice Modifier to restrict function access to accounts with a specific role.
+     * @param roleId ID of the required role.
+     */
+    modifier onlyRole(uint64 roleId) {
+        (bool hasRole, ) = accessManager.hasRole(roleId, msg.sender);
+        require(hasRole, 'LiquidityPool: sender must have the required role');
+        _;
+    }
```
[the remark this answers is not in the document]
Comment (MTthoas) [reply contracts/src/LiquidityPoolV2.sol:65]: Yeah nice see

File: contracts/src/LiquidityPoolFactoryV2.sol [hunk 12]
```diff
@@ -1,60 +1,87 @@
...
+    mapping(address => mapping(address => address)) public getPool;
 
-    // List of all deployed liquidity pools
+    // All address of pools
     address[] public allPools;
 
-    // Mapping from pair (tokenA, tokenB) to the liquidity pool address
-    mapping(address => mapping(address => address)) public getPool;
+    /** Initializer function (replaces constructor by OpenZeppelin standards) : https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable */
+    function initialize(address _accessManager) public override initializer {
```
[the remark this answers is not in the document]
Comment (MTthoas) [reply contracts/src/LiquidityPoolFactoryV2.sol:20]: Exactly, im using override cuz 1 function in : `lib/openzeppelin-contracts-upgradeable/contracts/access/manager/AccessManagerUpgradeable.sol:133:5:``is using initialize name.

File: contracts/src/LiquidityPoolV2.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (MTthoas) [reply contracts/src/LiquidityPoolV2.sol:65]: image

File: .gitmodules (hunk 8 above)
Comment (MTthoas) [reply .gitmodules:15]: its a bug, but the path works

**Pull request merged**